src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp

```diff
@@ -182,39 +184,73 @@ APPLICATION_INFO::EnsureApplicationCreated(
         return S_OK;
     }
 
-    if (m_pApplication == NULL)
+    if (m_fDoneAppCreation)
     {
-        SRWExclusiveLock lock(m_srwLock);
+        // application is NULL and CreateApplication failed previously
+        FINISHED_IF_FAILED(E_APPLICATION_ACTIVATION_EXEC_FAILURE);
+    }
+    else
+    {
+        AcquireSRWLockExclusive(&m_srwLock);
```

Comment:
Just create a SRWLockExclusive here so you can remove the fLocked bool

Reply:
I want make sure setting m_fDoneAppCreation= TRUE is inside the lock even for failure condition

Reply:
then put SRWLockExclusive  one level higher and it would be active until method exits.
I removed all manual locking from shim, please don't add it back.

Reply:
This also makes the logging part inside the lock. We need some flexibility to do the minimum lock.

Reply:
In your code lock would never be returned in case of an exception. We log under lock in a lot of places.